### ISSUE TITLE
Avoid blocking access to EntityManagerFactory in DataSourceInitializedPublisher

### ISSUE THREAD

**Issue by olivergierke** [spring-projects/spring-boot#14061]
During our optimization efforts in JPA bootstrapping we stumbled above `DataSourceInitializedPublisher.publishEventIfRequired(…)` calling `EntityManagerFactory.getProperties()`.

In case of an asynchronously bootstrapped `EntityManagerFactory` this will cause the initialization to block and prevent the container from continuing to initialize other components. I wonder it there was a way to rather consume the `DataSource` by inspecting `EntityManagerFactoryInfo.getDataSource()` as method calls to this interface do not trigger the blocking.

**Comment by wilkinsona** [comment]
The `DataSource` is used as part of determining whether or not a `DataSourceSchemaCreatedEvent` should be published. Such an event should be published when Hibernate is responsible for creating the database's schema and has done so. As such, the fact that the current implementation blocks until the `EntityManagerFactory` has complete its initialisation is helpful. If we switched to the non-blocking `EntityManagerFactoryInfo.getDataSource()` instead, the event would be published too soon.

It looks like we need a way of being notified that the `EntityManagerFactory` has been initialized and at that point when can publish the event. This may require a change in Framework.

**Comment by wilkinsona** [comment]
In my testing, the current call blocks for ~800ms when running on Data JPA sample. That's quite a large window in which some other processing could be done in parallel with the entity manager factory's initialisation. If we can figure out how to do it, the proposed change looks like it'll be beneficial.

**Comment by wilkinsona** [comment]
The code for these changes and for #13833 are available in [this branch](
https://github.com/wilkinsona/spring-boot/tree/gh-13833) of my fork.

Here are some numbers for starting the Data JPA sample on my laptop in Eclipse. Each entry is the average JVM running time of starting the app 10 times:

Default | Lazy | Deferred
 --- | ---- | ----
4.606 | 3.992 | 4.479

**Comment by olivergierke** [comment]
Looks good to me. We have seen the improvements to take even more effect rising with the number of entities in the system, i.e. the longer the EMF needs to build as that might allow even all application components to bootstrap just while the EMF finishes initialization.

Dave's (admittedly esotheric) sample app containing a couple of thousand entities and repositories could be brought down from > 70 seconds to just slightly over 10.

**Comment by sforner405** [comment]
By migrating to 2.3.5 we now having issues with MOCK tests.
webEnvironment = SpringBootTest.WebEnvironment.MOCK
leads to hanging DataSourceInitializedPublisher

**Comment by wilkinsona** [comment]
@sforner405 We've had a few problems in this area unfortunately but we hoped we'd straightened things out in 2.3.5. If that's not the case for you, can you please open a new issue and provide a small sample that reproduces the hang in your tests with a mock web environment?
